### PR TITLE
Remove history depth limit, keep only last visit per task in history, drop removed tasks from history

### PR DESCRIPTION
**Pull request by akuniutka**
Achieve sprint #6 goals:
- Remove visited tasks history depth limit for an in-memory task manager.
- Keep only the last visit for each task in a history with O(1) time complexity.
- Drop tasks, that are removed from a task manager, from a history manager, too.

To achieve O(1) time complexity in keeping the last visit per task in history, while saving the order they are visited, combin a hash map with a doubly linked list, where the hash map stores pairs "task id -> a list node containing the task".

Also, cover new features with tests and rearrenge existing tests:
- Move test model creation to a separate utility class.
- Move tests for correct history keeping to in-memory history manager tests.
- Aim In-memory task manager tests to checks of passing correct data to a history manager and back.

### CODE REVIEW

File: src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java [hunk 1]
```diff
@@ -3,26 +3,92 @@
...
 
 import java.util.ArrayList;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 
 public class InMemoryHistoryManager implements HistoryManager {
-    private static final int MAX_HISTORY_SIZE = 10;
-    private final List<Task> history;
+    private final Map<Long, Node> nodes;
```
Comment (YazykovMikhail) [inline src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java:11]: назови history, все таки историю хронят

File: src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java [hunk 2]
```diff
@@ -3,26 +3,92 @@
...
+    private Node tail;
 
     InMemoryHistoryManager() {
-        history = new ArrayList<>();
+        this.nodes = new HashMap<>();
     }
 
     @Override
     public void add(Task task) {
-        if (history.size() == MAX_HISTORY_SIZE) {
```
Comment (YazykovMikhail) [inline src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java:18]: Стоит в самом начале проверять на null
		if (task == null) {
			return;
		}
В друг задачу удалили и в историюпойдет null

File: src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java [hunk 3]
```diff
@@ -3,26 +3,92 @@
...
-            history.removeFirst();
+        final long id = task.getId();
+        final Node oldNode = nodes.get(id);
+        if (oldNode != null) {
+            removeNode(oldNode);
+        }
```
Comment (YazykovMikhail) [inline src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java:25]: Вместо 

```
        final Node oldNode = nodes.get(id);
        if (oldNode != null) {
            removeNode(oldNode);
        }
```
просто 		remove(id);

Comment (YazykovMikhail) [reply src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java:25]: Дальше покажу почему, немного уберем дулирование кода

File: src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java [hunk 4]
```diff
@@ -3,26 +3,92 @@
...
+        final Node newNode = linkLast(task);
+        nodes.put(id, newNode);
+    }
+
+    @Override
+    public void remove(long id) {
```
Comment (YazykovMikhail) [inline src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java:31]: public void remove(int id) {
		final Node node = history.remove(id);
		if (node == null) {
			return;
		}
		removeNode(node);
	}

File: src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java [hunk 5]
```diff
@@ -3,26 +3,92 @@
...
+        final ArrayList<Task> tasks = new ArrayList<>();
+        Node node = head;
+        while (node != null) {
+            tasks.add(node.task);
+            node = node.next;
+        }
+        return tasks;
+    }
+
+    private static class Node {
```
Comment (YazykovMikhail) [inline src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java:83]: Отлично

**Review by YazykovMikhail — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java (hunk 1 above)
Comment (akuniutka) [reply src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java]: Fixed

File: src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java (hunk 2 above)
Comment (akuniutka) [reply src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java:18]: Fixed

File: src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java (hunk 3 above)
Comment (akuniutka) [reply src/main/java/io/github/akuniutka/kanban/service/InMemoryHistoryManager.java:23]: Fixed

**Pull request merged**